features/cellular/framework/targets/RiotMicro/AT/RM1000_AT_CellularStack.cpp

```diff
@@ -0,0 +1,412 @@
+    bool success = true;
+    nsapi_size_t read_blk;
+    nsapi_size_t count = 0;
+    nsapi_size_t rsorcv_sz;
+    char ipAddress[NSAPI_IP_SIZE];
+    uint8_t ch = 0;
+    int port = 0;
+    Timer timer;
+
+    if (socket->pending_bytes == 0) {
+        _at.process_oob();
+        if (socket->pending_bytes == 0) {
+            return NSAPI_ERROR_WOULD_BLOCK;
+        }
+    }
+
+    timer.start();
+    while (success && (size > 0)) {
+        read_blk = RM1000_MAX_PACKET_SIZE;
+        if (read_blk > size) {
+            read_blk = size;
+        }
+        if (socket->pending_bytes > 0) {
+            _at.cmd_start("AT+RSORCV=");
+            _at.write_int(socket->id);
+            _at.write_int(read_blk);
+            _at.cmd_stop();
+
+            _at.resp_start("+RSORCV:");
+            _at.skip_param(); // receiving socket id
+            rsorcv_sz = _at.read_int();
+            if (socket->proto == NSAPI_UDP) {
+                _at.read_string(ipAddress, sizeof(ipAddress));
+                port = _at.read_int();
+            }
+            if (rsorcv_sz > size) {
+                rsorcv_sz = size;
+            }
+            _at.read_bytes((uint8_t *)buffer + count, rsorcv_sz);
+            _at.resp_stop();
+
+            // Must use what +RSORCV returns here as it may be less or more than we asked for
+            if (rsorcv_sz > socket->pending_bytes) {
+                socket->pending_bytes = 0;
+            } else {
+                socket->pending_bytes -= rsorcv_sz;
+            }
+
+            if (rsorcv_sz > 0) {
+                count += rsorcv_sz;
+                size -= rsorcv_sz;
+            } else {
+                // read() should not fail
+                success = false;
+            }
+        }  else if (timer.read_ms() < SOCKET_TIMEOUT) {
+            // Wait for URCs
+            _at.process_oob();
+        } else {
+            if (count == 0) {
+                // Timeout with nothing received
+                nsapi_error_size = NSAPI_ERROR_WOULD_BLOCK;
+                success = false;
+            }
+            size = 0; // This simply to cause an exit
+        }
+    }
+    timer.stop();
+
+    if (!count || (_at.get_last_error() != NSAPI_ERROR_OK)) {
+        return NSAPI_ERROR_WOULD_BLOCK;
+    } else {
+        nsapi_error_size = count;
+    }
+
+    if (success && socket->proto == NSAPI_UDP && address) {
+        address->set_ip_address(ipAddress);
+        address->get_ip_address();
+        address->set_port(port);
+    }
+
+    return nsapi_error_size;
+}
+
+nsapi_error_t RM1000_AT_CellularStack::socket_close_impl(int sock_id)
+{
+    tr_debug("RM1000_AT_CellularStack::socket_close_impl");
+
+    _at.lock();
+    _at.cmd_start("AT+RSOCL=");
+    _at.write_int(sock_id);
+    _at.cmd_stop_read_resp();
+
+    return _at.unlock_return_error();
+}
+
+// Find or create a socket from the list.
+RM1000_AT_CellularStack::CellularSocket *RM1000_AT_CellularStack::find_socket(int id)
```

Comment:
Can this be removed? It looks the same as `AT_CellularStack::find_socket`.

Reply:
removed